src/Mixture.cc

```diff
@@ -226,36 +226,70 @@ void Mixture::evaluateComponents(
     }
 }
 
+void Mixture::evaluateDerivatives(
+    Eigen::Matrix<Scalar, Eigen::Dynamic, 1> & x,
+    Eigen::Matrix<Scalar, Eigen::Dynamic, 1> & gradient,
+    Eigen::Matrix<Scalar, Eigen::Dynamic, Eigen::Dynamic> & hessian
+) const {
+    evaluateDerivativesImpl(x, gradient, &hessian);
+}
+
+void Mixture::evaluateDerivatives(
+    Eigen::Matrix<Scalar, Eigen::Dynamic, 1> & x,
+    Eigen::Matrix<Scalar, Eigen::Dynamic, 1> & gradient
+) const {
+    evaluateDerivativesImpl<Eigen::Matrix<Scalar, Eigen::Dynamic, 1>,
+                            Eigen::Matrix<Scalar, Eigen::Dynamic, 1>,
+                            Eigen::Matrix<Scalar, Eigen::Dynamic, Eigen::Dynamic>> (x, gradient, nullptr, false);
+}
+
 void Mixture::evaluateDerivatives(
     ndarray::Array<Scalar const,1,1> const & x,
     ndarray::Array<Scalar,1,1> const & gradient,
     ndarray::Array<Scalar,2,1> const & hessian
+) const {
+    auto hessianEigen = ndarray::asEigenMatrix(hessian);
+    auto gradientEigen = ndarray::asEigenMatrix(gradient);
+    evaluateDerivativesImpl(ndarray::asEigenMatrix(x), gradientEigen, &hessianEigen);
+}
+
+template <typename A, typename B, typename C>
+void Mixture::evaluateDerivativesImpl(
+    A const & x,
+    B & gradient,
+    C * hessian,
+    bool computeHessian
 ) const {
     LSST_THROW_IF_NE(
-        x.getSize<0>(), _dim,
+        x.size(), _dim,
         pex::exceptions::LengthError,
         "Size of x array (%d) does not dimension of mixture (%d)"
     );
     LSST_THROW_IF_NE(
-        gradient.getSize<0>(), _dim,
+        gradient.rows(), _dim,
         pex::exceptions::LengthError,
         "Size of gradient array (%d) does not dimension of mixture (%d)"
     );
-    LSST_THROW_IF_NE(
-        hessian.getSize<0>(), _dim,
-        pex::exceptions::LengthError,
-        "Number of rows of hessian array (%d) does not dimension of mixture (%d)"
-    );
-    LSST_THROW_IF_NE(
-        hessian.getSize<1>(), _dim,
-        pex::exceptions::LengthError,
-        "Number of columns of hessian array (%d) does not dimension of mixture (%d)"
-    );
-    gradient.deep() = 0.0;
-    hessian.deep() = 0.0;
+    gradient.setZero();
+
+    if (computeHessian) {
```

Comment:
Might want to add an `assert(hessian)` here if internal logic is supposed to guarantee that it's always nonnull when `computeHessian` is true, or an exception-throwing check if there's no internal guarantee on that.